src/main/java/jenkins/branch/MultiBranchProject.java

```diff
@@ -358,6 +362,149 @@ public synchronized void setProjectFactory(BranchProjectFactory<P, R> projectFac
+            // either adding, removing, or updating without an id change
+            this.sources.replaceBy(sources);
+            return;
+        }
+        // Now we need to check if any of the new entries are effectively the same as an old entry that is being removed
+        // we will store the ID changes in a map and process all the affected branches to update their sourceIds
+        Map<String,String> changedIds = new HashMap<>();
+        Set<String> additions = new HashSet<>(newIds);
+        additions.removeAll(oldIds);
+        Set<String> removals = new HashSet<>(oldIds);
+        removals.removeAll(newIds);
+
+        // we will be manipulating the sources, so make sure we own the backing list impl
+        sources = new ArrayList<>(sources);
+        for (ListIterator<BranchSource> i = sources.listIterator(); i.hasNext(); ) {
+            BranchSource addition = i.next();
+            String additionId = addition.getSource().getId();
+            if (!additions.contains(additionId)) {
+                continue;
+            }
+            for (BranchSource removal: this.sources) {
+                String removalId = removal.getSource().getId();
+                if (!removals.contains(removalId)) {
+                    continue;
+                }
+                if (!equalButForId(removal.getSource(), addition.getSource())) {
+                    continue;
+                }
+                changedIds.put(removalId, additionId);
+                // now take these two out of consideration
+                removals.remove(removalId);
+                additions.remove(additionId);
+                break;
+            }
+        }
+        this.sources.replaceBy(sources);
+        BranchProjectFactory<P,R> factory = getProjectFactory();
+        for (P item: getItems()) {
+            if (!factory.isProject(item)) {
+                continue;
+            }
+            Branch oldBranch = factory.getBranch(item);
+            if (changedIds.containsKey(oldBranch.getSourceId())) {
+                Branch newBranch = new Branch(changedIds.get(oldBranch.getSourceId()), oldBranch.getHead(), oldBranch.getScm(), oldBranch.getProperties());
+                newBranch.setActions(oldBranch.getActions());
+                factory.setBranch(item, newBranch);
+            }
+        }
+    }
+
+    private Set<String> sourceIds(List<BranchSource> sources) {
+        Set<String> result = new HashSet<>();
+        for (BranchSource s: sources) {
+            result.add(s.getSource().getId());
+        }
+        return result;
+    }
+
+    private boolean equalButForId(SCMSource a, SCMSource b) {
+        if (!a.getClass().equals(b.getClass())) {
+            return false;
+        }
+        boolean fallBack = false;
+        Class<?> clazz = a.getClass();
+        REFLECTION:
+        while (clazz != null && SCMSource.class.isAssignableFrom(clazz) && clazz != SCMSource.class) {
+            for (Field field : clazz.getDeclaredFields()) {
```

Comment:
Yep. It's premature optimisation... from before I found `omitField` and was doing crazy things to ensure a bare ID would end up escaped the same way as in the whole object... I guess I'll put the reflection code back in my rainy day library